### PR TITLE
feat(client): add WithZones and WithRegions options

### PR DESCRIPTION
**Pull request by Codelax**
Complete region
```go
servers, err := api.ListServers(&instance.ListServersRequest{},
	scw.WithZones(scw.RegionFrPar.GetZones()...),
)
```

Specific zones
```go
servers, err := api.ListServers(&instance.ListServersRequest{},
	scw.WithZones(scw.ZoneFrPar1, scw.ZoneFrPar2, scw.ZoneNlAms1),
)
```

All zones
```go
servers, err := api.ListServers(&instance.ListServersRequest{},
	scw.WithZones(scw.AllZones...),
)
```

### CODE REVIEW

File: scw/client.go [hunk 1]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+		}
+
+		// Requests are parallelized
+		responseMutex := sync.Mutex{}
+		requestGroup := sync.WaitGroup{}
+		errChan := make(chan error, len(zones))
+
+		requestGroup.Add(len(zones))
+		for _, zone := range zones {
+			go func(zone Zone) {
```
Comment (Monitob) [inline scw/client.go:374]: Can you put this in a named function

File: scw/client.go [hunk 2]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+				_, err = response.UnsafeAppend(zoneResponse)
+				responseMutex.Unlock()
+				if err != nil {
+					errChan <- err
+				}
+			}(zone)
+		}
+		requestGroup.Wait()
+
+	L: // We gather potential errors and return them all together
```
Comment (remyleone) [inline scw/client.go:398]: Is there a way to avoid this goto syntax?

File: scw/client.go [hunk 3]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+		for {
+			select {
+			case newErr := <-errChan:
+				err = errors.Wrap(err, newErr.Error())
```
Comment (Monitob) [inline scw/client.go:402]: Can we put this on a slide ?

File: scw/client.go [hunk 4]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+
+		sortResponseByZone(res)
+		return nil
+	}
+
+	return errors.New("%T does not support pagination", res)
+}
+
+// sortSliceByZone sorts a slice of struct using a Zone field that should exist
+func sortSliceByZone(list interface{}) {
```
Comment (remyleone) [inline scw/client.go:420]: Is there a way to sort them in the order they are passed? if a user ask for nl-ams-1, fr-par-2 and fr-par-1 could we follow this order?

File: scw/client.go [hunk 5]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
 	return errors.New("%T does not support pagination", res)
 }
 
+// doListZones collects all zones using multiple list requests and aggregate all results on a single response.
+// result is sorted by zone
+func (c *Client) doListZones(req *ScalewayRequest, res interface{}, zones []Zone) (err error) {
```
Comment (Monitob) [inline scw/client.go:353]: What about to return two channels (res, errors)

File: scw/client.go [hunk 6]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+
+		// Requests are parallelized
+		responseMutex := sync.Mutex{}
+		requestGroup := sync.WaitGroup{}
+		errChan := make(chan error, len(zones))
+
+		requestGroup.Add(len(zones))
+		for _, zone := range zones {
+			go func(zone Zone) {
+				defer requestGroup.Done()
```
Comment (Monitob) [inline scw/client.go:375]: What happen when one context fail?

File: scw/client.go (hunk 5 above)
Comment (Monitob) [inline scw/client.go:353]: Can we also handle the context present on the client

Comment (Monitob) [reply scw/client.go:353]: Also we can wrap all the errors on one

File: scw/client.go [hunk 7]
```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
...
+	// type ListServersResponse struct {
+	//	TotalCount uint32 `json:"total_count"`
+	//	Servers []*Server `json:"servers"`
+	// }
+	// We iterate over fields searching for the slice one to sort it
+	resType := reflect.TypeOf(res).Elem()
+	fields := reflect.VisibleFields(resType)
+	for _, field := range fields {
+		if field.Type.Kind() == reflect.Slice {
+			sortSliceByZone(reflect.ValueOf(res).Elem().FieldByName(field.Name).Interface())
```
Comment (Monitob) [inline scw/client.go:442]: We should respect the sort asked by the user

File: scw/client.go (hunk 5 above)
Comment (Monitob) [reply scw/client.go:353]: Only one context is ok

File: scw/client.go (hunk 2 above)
Comment (Codelax) [reply scw/client.go]: break stop current `select`, label is needed to stop `for`

File: scw/client.go (hunk 7 above)
Comment (Codelax) [reply scw/client.go:503]: Fixed by using stable sort, user's sort will be respected but result will still be sorted by locality

File: scw/client.go [hunk 8]
```diff
@@ -340,6 +351,182 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
 	return errors.New("%T does not support pagination", res)
 }
 
+// doListLocalities collects all localities using mutliple list requests and aggregate all results on a lister response
+// results is sorted by locality
+func (c *Client) doListLocalities(req *ScalewayRequest, res lister, localities []string) (err error) {
```
Comment (Monitob) [inline scw/client.go:356]: Can you add an example to the Getting Stated Readme

Comment (Monitob) [reply scw/client.go:356]: Please check the examples here https://github.com/scaleway/scaleway-sdk-go/blob/v1.0.0-beta.9/example_test.go#L160

**Review by Monitob — APPROVED** [review]

**Pull request merged**